man/hfst-strings2fst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-STRINGS2FST "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-strings2fst \- =Compile string pairs and pair-strings into transducer(s)
.SH SYNOPSIS
.B hfst-strings2fst
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Compile string pairs and pair\-strings into transducer(s)
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input strings from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "String and format options:"
.TP
\fB\-f\fR, \fB\-\-format\fR=\fI\,FMT\/\fR
Write result in FMT format
.TP
\fB\-j\fR, \fB\-\-disjunct\-strings\fR
Disjunct all strings instead of transforming
each string into a separate transducer
.TP
\fB\-\-norm\fR
Divide each weight by sum of all weights
(with option \fB\-j\fR)
.TP
\fB\-\-log\fR
Take negative natural logarithm of each weight
.TP
\fB\-\-log10\fR
Take negative 10\-based logarithm of each weight
.TP
\fB\-p\fR, \fB\-\-pairstrings\fR
Input is in pairstring format
.TP
\fB\-S\fR, \fB\-\-has\-spaces\fR
Input has spaces between symbols/symbol pairs
.TP
\fB\-e\fR, \fB\-\-epsilon\fR=\fI\,EPS\/\fR
Interpret string EPS as epsilon.
.TP
\fB\-m\fR, \fB\-\-multichar\-symbols\fR=\fI\,FILE\/\fR
Strings that must be tokenized as one symbol.
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
FMT can be { foma, openfst\-tropical, openfst\-log, sfst,
optimized\-lookup\-weighted, optimized\-lookup\-unweighted }.
If EPS is not defined, the default representation of @0@ is used.
Option \fB\-\-norm\fR precedes option \fB\-\-log\fR.
The FILE of option \fB\-m\fR lists all multichar\-symbols, each symbol
on its own line.
Backslash '\e' may be used to escape ':', tab and itself. For any
other symbol x '\ex' means x literally, i.e. is the same as 'x'.
The weight of a string can be given after the string separated
by a tabulator. The weight cannot be zero.
.SH EXAMPLES
.TP
echo "cat:dog" | hfst\-strings2fst
create cat:dog fst
.TP
echo "c:da:ot:g" | hfst\-strings2fst \-p
same as pairstring
.TP
echo "c:d a:o t:g" | hfst\-strings2fst \-p \-S
same as pairstring with spaces
.TP
echo "c a t:d o g" | hfst\-strings2fst \-S
same with spaces
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-strings2fst home page:
<https://github.com/hfst/hfst/wiki/Strings2Fst>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-strings2fst.cc
//! @file hfst-strings2fst.cc
//!
//! @brief string compiling command line tool
//!
//! @author HFST Team

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif

#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <vector>

using std::pair;
using std::vector;

#include <cstdio>
#include <cstdlib>
#include <cstring>

#ifdef _MSC_VER
#include "hfst-getopt.h"
#else
#include <getopt.h>
#endif

#include <errno.h>
#include <math.h>

#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "HfstStrings2FstTokenizer.h"
#include "HfstTransducer.h"
#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "implementations/HfstBasicTransducer.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstOutputStream;
using hfst::HfstTokenizer;
using hfst::HfstTransducer;
using hfst::implementations::HfstBasicTransducer;
// using hfst::HfstInternalTransducer;
// using hfst::implementations::HfstTrie;
using hfst::StringPair;
using hfst::StringPairVector;
using hfst::StringSet;
using hfst::StringVector;

static char *epsilonname = NULL; // FIX: use this
static bool has_spaces = false;
static bool disjunct_strings = false;
static bool pairstrings = false;
static char *multichar_symbol_filename = NULL;
static StringVector multichar_symbols;

static float sum_of_weights = 0;
static bool normalize_weights = false;
static bool logarithmic_weights_e = false;
static bool logarithmic_weights_10 = false;

static bool warn_negative_weights = true;
static bool warnings_are_errors = false;

static hfst::ImplementationType output_format = hfst::UNSPECIFIED_TYPE;

float
divide_by_sum_of_weights(float weight)
{
    if (sum_of_weights == 0)
        return 0;
    return weight / sum_of_weights;
}
float
take_negative_logarithm_e(float weight)
{
    float result;
    if (weight == 0)
        result = 0; // shoud be INFINITY, but doesn't work in transitions
    else
    {
        errno = 0;
        result = -log(weight);
        if (errno != 0)
        {
            error(EXIT_FAILURE, errno, "unable to take negative logarithm");
        }
    }
    return result;
}
float
take_negative_logarithm_10(float weight)
{
    float result;
    if (weight == 0)
        result = 0; // shoud be INFINITY, but doesn't work in transitions
    else
    {
        errno = 0;
        result = -log10(weight);
        if (errno != 0)
        {
            error(EXIT_FAILURE, errno, "unable to take negative logarithm");
        }
    }
    return result;
}

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out,
            "Usage: %s [OPTIONS...] [INFILE]\n"
            "Compile string pairs and pair-strings into transducer(s)\n"
            "\n",
            program_name);
    print_common_program_options(message_out);
    fprintf(message_out,
            "Input/Output options:\n"
            "  -i, --input=INFILE     Read input strings from INFILE\n"
            "  -o, --output=OUTFILE   Write output transducer to OUTFILE\n");
    fprintf(
        message_out,
        "String and format options:\n"
        "  -f, --format=FMT          Write result in FMT format\n"
        "  -j, --disjunct-strings    Disjunct all strings instead of "
        "transforming\n"
        "                            each string into a separate transducer\n"
        /*"      --sum                 Sum weights of duplicate strings\n"
          "                            instead of taking minimum\n"*/
        "      --norm                Divide each weight by sum of all "
        "weights\n"
        "                            (with option -j)\n"
        "      --log                 Take negative natural logarithm of each "
        "weight\n"
        "      --log10               Take negative 10-based logarithm of each "
        "weight\n"
        "  -p, --pairstrings         Input is in pairstring format\n"
        "  -S, --has-spaces          Input has spaces between symbols/symbol "
        "pairs\n"
        "  -e, --epsilon=EPS         Interpret string EPS as epsilon.\n"
        "  -m, --multichar-symbols=FILE   Strings that must be tokenized as "
        "one symbol.\n");
    fprintf(message_out, "\n");

    fprintf(
        message_out,
        "If OUTFILE or INFILE is missing or -, standard streams will be "
        "used.\n"
        "FMT can be { foma, openfst-tropical, openfst-log, sfst, \n"
        "optimized-lookup-weighted, optimized-lookup-unweighted }.\n"
        "If EPS is not defined, the default representation of @0@ is used.\n"
        "Option --norm precedes option --log.\n"
        "The FILE of option -m lists all multichar-symbols, each symbol\n"
        "on its own line.\n"
        "Backslash '\\' may be used to escape ':', tab and itself. For any\n"
        "other symbol x '\\x' means x literally, i.e. is the same as 'x'.\n"
        "The weight of a string can be given after the string separated\n"
        "by a tabulator. The weight cannot be zero.\n"
        "\n");

    fprintf(
        message_out,
        "Examples:\n"
        "  echo \"cat:dog\" | %s            create cat:dog fst\n"
        "  echo \"c:da:ot:g\" | %s -p       same as pairstring\n"
        "  echo \"c:d a:o t:g\" | %s -p -S  same as pairstring with spaces\n"
        "  echo \"c a t:d o g\" | %s -S     same with spaces\n"
        "\n",
        program_name, program_name, program_name, program_name);
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
    fprintf(message_out, "\n");
}

int
parse_options(int argc, char **argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[]
            = { HFST_GETOPT_COMMON_LONG,
                HFST_GETOPT_UNARY_LONG,
                { "disjunct-strings", no_argument, 0, 'j' },
                { "epsilon", required_argument, 0, 'e' },
                { "norm", no_argument, 0, '2' },
                { "log", no_argument, 0, '3' },
                { "log10", no_argument, 0, '4' },
                { "pairstrings", no_argument, 0, 'p' },
                { "has-spaces", no_argument, 0, 'S' },
                { "multichar-symbols", required_argument, 0, 'm' },
                { "format", required_argument, 0, 'f' },
                { "Wstuff", required_argument, 0, 'W' },
                { 0, 0, 0, 0 } };
        int option_index = 0;
        int c = getopt_long(argc, argv,
                            HFST_GETOPT_COMMON_SHORT HFST_GETOPT_UNARY_SHORT
                            "je:234pSm:f:W:",
                            long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'e':
            epsilonname = hfst_strdup(optarg);
            break;
        case '2':
            normalize_weights = true;
            break;
        case '3':
            logarithmic_weights_e = true;
            break;
        case '4':
            logarithmic_weights_10 = true;
            break;
        case 'j':
            disjunct_strings = true;
            break;
        case 'S':
            has_spaces = true;
            break;
        case 'p':
            pairstrings = true;
            break;
        case 'm':
            multichar_symbol_filename = hfst_strdup(optarg);
            break;
        case 'f':
            output_format = hfst_parse_format_name(optarg);
            break;
        case 'W':
            if (strcmp(optarg, "error") == 0)
            {
                warnings_are_errors = true;
            }
            else if (strcmp(optarg, "no-error") == 0)
            {
                warnings_are_errors = false;
            }
            else if (strcmp(optarg, "negative-weights") == 0)
            {
                warn_negative_weights = true;
            }
            else if (strcmp(optarg, "no-negative-weights") == 0)
            {
                warn_negative_weights = false;
            }
            else
            {
                char *errm = (char *)malloc(sizeof(char) * 64);
                sprintf(errm, "unrecognised warning option -W%s", optarg);
                hfst_error(EXIT_FAILURE, 0, errm);
                return EXIT_FAILURE;
            }
            break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    if (output_format == hfst::UNSPECIFIED_TYPE)
    {
        verbose_printf("Output format not specified, "
                       "defaulting to openfst tropical\n");
        output_format = hfst::TROPICAL_OPENFST_TYPE;
    }
    if (epsilonname == 0)
    {
        epsilonname = hfst_strdup("@0@");
    }
    return EXIT_CONTINUE;
}

int
process_stream(HfstOutputStream &outstream)
{
    size_t transducer_n = 0;
    char *line = 0;
    size_t len = 0;
    HfstTokenizer tok;
    HfstBasicTransducer disjunction;
    size_t line_n = 0;

    hfst::HfstStrings2FstTokenizer multichar_symbol_tokenizer(
        multichar_symbols, std::string(epsilonname));

    while (hfst_getline(&line, &len, inputfile) != -1)
    {
        transducer_n++;
        line_n++;
        verbose_printf("Parsing line %u...\n", line_n);
        // parse line end and weight
        char *tab = strstr(line, "\t");
        char *string_end = tab;
        double weight = 0.0;
        bool weighted = false;
        if (tab == NULL)
        {
            string_end = line;
            while ((*string_end != '\0') && (*string_end != '\n')
                   && (*string_end != '\r'))
            {
                string_end++;
            }
        }
        else
        {
            // change '\n' to '\0'
            char *p = tab;
            while (*p != '\0')
            {
                if ((*p == '\n') || (*p == '\r'))
                {
                    *p = '\0';
                }
                p++;
            }

            weight = hfst_strtoweight(tab + 1);
            weighted = true;
            char *errm = (char *)malloc(sizeof(char) * 256);
            sprintf(errm,
                    "Found negative weight %f; negative weights are "
                    "supported but iffy, if you really need them use "
                    "-Wno-negative-weights",
                    weight);
            if ((weight < 0) && warn_negative_weights)
            {
                if (warnings_are_errors)
                {
                    hfst_error_at_line(EXIT_FAILURE, 0, inputfilename, line_n,
                                       errm);
                }
                else
                {
                    hfst_warning_at_line(0, 0, inputfilename, line_n, errm);
                }
            }
        }
        *string_end = '\0';
        // Parse the string
        StringPairVector spv;
        try
        {
            if (pairstrings)
            {
                spv = multichar_symbol_tokenizer.tokenize_pair_string(
                    line, has_spaces);
            }
            else
            {
                spv = multichar_symbol_tokenizer.tokenize_string_pair(
                    line, has_spaces);
            }
        }
        catch (const hfst::UnescapedColsFound &e)
        {
            if (pairstrings)
            {
                error_at_line(EXIT_FAILURE, errno, inputfilename, line_n,
                              "String `%s' contains unescaped ':'-symbols,\n"
                              "which are not pair separators. Use `\\:' for "
                              "literal `:'.",
                              line);
            }
            else
            {
                error_at_line(
                    EXIT_FAILURE, errno, inputfilename, line_n,
                    "String `%s' contains unescaped `:'-symbols,\n"
                    "which are not pair separators. Use `\\:\' for "
                    "literal "
                    "`:'.\n"
                    "If you are compiling pair strings, use option -p.",
                    line);
            }
        }
        catch (const IncorrectUtf8CodingException &e)
        {
            error_at_line(EXIT_FAILURE, errno, inputfilename, line_n,
                          "Input string `%s' is not valid utf-8.", line);
        }
        // Handle the weight
        float path_weight = 0;

        if (weighted)
        {
            sum_of_weights = sum_of_weights + weight;
            path_weight = weight;
            verbose_printf("Using final weight %f...\n", weight);
        }

        if (!disjunct_strings) // each string into a transducer
        {
            HfstBasicTransducer tr;

            if (logarithmic_weights_e)
            {
                path_weight = take_negative_logarithm_e(weight);
            }
            else if (logarithmic_weights_10)
            {
                path_weight = take_negative_logarithm_10(weight);
            }
            else
            {
            }

            tr.disjunct(spv, path_weight);
            HfstTransducer res(tr, output_format);
            hfst_set_name(res, "", "string");
            outstream << res;
        }
        else // disjunct all strings into a single transducer
        {
            // do not take negative logarithm yet
            disjunction.disjunct(spv, path_weight);
        }
    }
    free(line);
    if (disjunct_strings)
    {
        HfstTransducer res(disjunction, output_format);

        if (normalize_weights)
        {
            verbose_printf("Normalising weights...\n");
            res.transform_weights(&divide_by_sum_of_weights);
        }
        if (logarithmic_weights_e)
        {
            verbose_printf("Taking negative logarithm...\n");
            res.transform_weights(&take_negative_logarithm_e);
        }
        else if (logarithmic_weights_10)
        {
            verbose_printf("Taking negative logarithm...\n");
            res.transform_weights(&take_negative_logarithm_10);
        }
        else
        {
        }

        hfst_set_name(res, "?", "strings");
        outstream << res;
    }
    return EXIT_SUCCESS;
}

int
main(int argc, char **argv)
{
#ifdef WINDOWS
    _setmode(1, _O_BINARY);
#endif

    hfst_set_program_name(argv[0], "0.1", "Strings2Fst");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }

    if (multichar_symbol_filename != NULL)
    {
        verbose_printf("Reading multichar symbols from %s\n",
                       multichar_symbol_filename);
        std::ifstream multichar_in(multichar_symbol_filename);
        (void)multichar_in.peek();
        if (!multichar_in.good())
        {
            error(EXIT_FAILURE, errno, "Multichar symbol file can't be read.");
        }
        char multichar_line[1000];
        while (multichar_in.good())
        {
            multichar_in.getline(multichar_line, 1000);
            if (strlen(multichar_line) > 0)
            {
                verbose_printf("Defining multichar symbol %s\n",
                               multichar_line);
                multichar_symbols.push_back(multichar_line);
            }
        }
    }

    // close output buffers, we use output streams
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n", inputfilename,
                   outfilename);
    // here starts the buffer handling part
    auto outstream
        = (outfile != stdout)
              ? std::make_unique<HfstOutputStream>(outfilename, output_format)
              : std::make_unique<HfstOutputStream>(output_format);
    process_stream(*outstream);
    free(inputfilename);
    free(outfilename);
    return EXIT_SUCCESS;
}
